dlib/test/serialize.cpp

```diff
@@ -401,18 +401,52 @@ namespace
         dlib::deserialize(item.b_false,in);
     }
     
+    template<typename T>
+    bool pointers_equal(const unique_ptr<T>& l, const unique_ptr<T>& r)
+    {
+        bool ptrs_equal = l.operator bool() == r.operator bool();
+        if (ptrs_equal && l)
+            ptrs_equal &= (*l == *r);
```

Comment:
This is doing a bitwise operation on a bool, which is odd and I don't know if the standard actually guarantees this is portable.

And don't call `.operator bool()` directly.  Write something like
```
if (l && r) 
   return *l == *r;
else
   return l == r;
```

Reply:
```l  == r``` will check if the underlying pointers are equal according to cppreference. We don't want that surely.

Reply:
I shouldn't have used the variable name ```ptrs_equal```. I should have used ```ptr_values_equal```

Reply:
If the pointer values are equal then the values are equal too.

Reply:
Oh i see. Thank you